File: Examples/ocaml/basics.ml
(* OCaml の基本 *)

let rec fib n =
  if n < 2 then n
  else fib (n - 1) + fib (n - 2)

let square x = x * x

let rec sum lst =
  match lst with
  | [] -> 0
  | h :: t -> h + sum t

type shape =
  | Circle of float
  | Rect of float * float

let area s =
  match s with
  | Circle r -> 3.14159 *. r *. r
  | Rect (w, h) -> w *. h

let classify n =
  if n < 0 then "negative"
  else if n = 0 then "zero"
  else if n < 10 then "small"
  else "large"

let () =
  let fibs = List.map fib [0; 1; 2; 3; 4; 5; 6; 7; 8; 9] in
  print_endline ("fib: " ^ String.concat ", " (List.map string_of_int fibs));

  let nums = [5; 3; 9; 1; 7] in
  Printf.printf "sum: %d\n" (sum nums);
  Printf.printf "fold: %d\n" (List.fold_left (fun a b -> a + b) 0 nums);
  print_endline ("squares: " ^ String.concat ", "
                   (List.map string_of_int (List.map square nums)));
  print_endline ("big: " ^ String.concat ", "
                   (List.map string_of_int (List.filter (fun x -> x > 3) nums)));
  print_endline ("sorted: " ^ String.concat ", "
                   (List.map string_of_int (List.sort compare nums)));
  Printf.printf "length: %d\n" (List.length nums);

  Printf.printf "circle area: %.3f\n" (area (Circle 2.0));
  Printf.printf "rect area: %.3f\n" (area (Rect (3.0, 4.0)));

  List.iter (fun n -> Printf.printf "%d is %s\n" n (classify n)) [-5; 0; 3; 42];

  let point = (3, 4) in
  let (x, y) = point in
  Printf.printf "point: %d, %d\n" x y;

  let add a b = a + b in
  let add5 = add 5 in
  Printf.printf "curried: %d\n" (add5 10);

  let text = "Hello, OCaml" in
  Printf.printf "upper: %s\n" (String.uppercase_ascii text);
  Printf.printf "length: %d\n" (String.length text);
  Printf.printf "sub: %s\n" (String.sub text 0 5);

  Printf.printf "17 mod 5 = %d\n" (17 mod 5);
  Printf.printf "concat: %s\n" (String.concat "-" ["a"; "b"; "c"]);
  print_endline (match 7 with 1 -> "one" | 7 -> "seven" | _ -> "other")

File: Examples/haskell/basics.hs
-- Haskell の基本
module Main where

fib :: Int -> Int
fib 0 = 0
fib 1 = 1
fib n = fib (n - 1) + fib (n - 2)

square :: Int -> Int
square x = x * x

mySum :: [Int] -> Int
mySum [] = 0
mySum (x:xs) = x + mySum xs

classify :: Int -> String
classify n
  | n < 0 = "negative"
  | n == 0 = "zero"
  | n < 10 = "small"
  | otherwise = "large"

data Shape = Circle Double | Rect Double Double

area :: Shape -> Double
area s = case s of
  Circle r -> 3.14159 * r * r
  Rect w h -> w * h

describe :: Shape -> String
describe (Circle r) = "circle of radius " ++ show r
describe (Rect w h) = "rect " ++ show w ++ "x" ++ show h

applyTwice :: (Int -> Int) -> Int -> Int
applyTwice f x = f (f x)

main :: IO ()
main = do
  let fibs = map fib [0..9]
  putStrLn ("fib: " ++ show fibs)

  let nums = [5, 3, 9, 1, 7]
  putStrLn ("sum: " ++ show (mySum nums))
  putStrLn ("sum': " ++ show (sum nums))
  putStrLn ("squares: " ++ show (map square nums))
  putStrLn ("big: " ++ show (filter (> 3) nums))
  putStrLn ("sorted: " ++ show (sort nums))
  putStrLn ("reversed: " ++ show (reverse nums))
  putStrLn ("length: " ++ show (length nums))
  putStrLn ("folded: " ++ show (foldl (+) 0 nums))

  mapM_ (\n -> putStrLn (show n ++ " is " ++ classify n)) [-5, 0, 3, 42]

  putStrLn (describe (Circle 2.0))
  putStrLn (describe (Rect 3.0 4.0))
  putStrLn ("area: " ++ show (area (Circle 2.0)))

  let doubled = [x * 2 | x <- nums, x > 3]
  putStrLn ("comprehension: " ++ show doubled)

  putStrLn ("applyTwice: " ++ show (applyTwice square 3))
  putStrLn ("curried: " ++ show (map (+ 1) nums))

  let (a, b) = (3, 4)
  putStrLn ("tuple: " ++ show a ++ "," ++ show b)

  putStrLn ("upper: " ++ toUpper "hello haskell")
  putStrLn ("words: " ++ show (words "a b c"))
  putStrLn ("17 div 5 = " ++ show (div 17 5))
  putStrLn ("17 mod 5 = " ++ show (mod 17 5))
  putStrLn ("zip: " ++ show (zipWith (+) [1, 2, 3] [10, 20, 30]))
